File: flutter/packages/hatake_core/lib/src/definition/chart_definition.dart
import 'package:equatable/equatable.dart';

import 'chart_kinds.dart';

/// How a `chart` item plots the rows it fetched.
///
/// Each point takes its label from [labelField] and its value from
/// [valueField]. With [aggregate] set, rows sharing a label are folded into one
/// point (see `AggregateRegistry.aggregateBy`); without it every row is a point,
/// which is what a pre-aggregated endpoint wants.
class ChartDefinition extends Equatable {
  /// Chart kind ([ChartKinds] or a plugin's).
  final String kind;

  /// Field holding each point's label (the category axis).
  final String labelField;

  /// Field holding each point's value. Not needed when [aggregate] is `count`.
  final String? valueField;

  /// Aggregate operation applied per label, or null to plot rows as they are.
  final String? aggregate;

  const ChartDefinition({
    this.kind = ChartKinds.bar,
    required this.labelField,
    this.valueField,
    this.aggregate,
  });

  @override
  List<Object?> get props => [kind, labelField, valueField, aggregate];
}

File: java/src/main/java/io/hatake/core/ChartDefinition.java
package io.hatake.core;

/**
 * ダッシュボードの {@code chart} カードが、取得した行を点の列にする方法。
 *
 * <p>{@code aggregate} があるとラベルが同じ行を1点に畳む（{@link Aggregates#aggregateBy}）。
 * 無ければ1行＝1点で、集計済みのエンドポイントをそのまま描くときの形。
 *
 * @param kind チャート種別（{@code bar} / {@code line} / {@code pie} or プラグイン）
 * @param labelField 各点のラベルを持つ項目名
 * @param valueField 各点の値を持つ項目名（{@code count} では不要）
 * @param aggregate ラベル別に適用する集約。null なら行をそのまま点にする
 */
public record ChartDefinition(
        String kind,
        String labelField,
        String valueField,
        String aggregate) {
}
